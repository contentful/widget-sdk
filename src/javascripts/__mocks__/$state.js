export default {
  go: jest.fn(),
  href: jest.fn().mockImplementation(sref => `http://url-for-state-${sref}`)
};
