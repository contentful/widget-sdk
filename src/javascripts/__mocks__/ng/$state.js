export const go = jest.fn();
export const href = jest.fn().mockImplementation(sref => `http://url-for-state-${sref}`);

export default {
  go,
  href
};
