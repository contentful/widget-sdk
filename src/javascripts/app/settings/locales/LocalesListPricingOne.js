import React from 'react';
import PropTypes from 'prop-types';
import Icon from 'ui/Components/Icon';
import { Workbench } from '@contentful/forma-36-react-components/dist/alpha';
import _ from 'lodash';
import { Button, Notification, Heading } from '@contentful/forma-36-react-components';
import KnowledgeBase from 'components/shared/knowledge_base_icon/KnowledgeBase';
import LocalesTable from './LocalesTable';
import StateLink from 'app/common/StateLink';
import LocalesUsageStatus, { getLocalesUsageStatus } from './utils/LocalesUsageStatus';

const LocalesListPropTypes = {
  locales: PropTypes.arrayOf(PropTypes.object).isRequired,
  canCreateMultipleLocales: PropTypes.bool.isRequired,
  canChangeSpace: PropTypes.bool.isRequired,
  localeResource: PropTypes.object.isRequired,
  insideMasterEnv: PropTypes.bool.isRequired,
  subscriptionState: PropTypes.object,
  subscriptionPlanName: PropTypes.string.isRequired,
  getComputeLocalesUsageForOrganization: PropTypes.func.isRequired
};

export const AddLocaleButton = ({ getComputeLocalesUsageForOrganization }) => (
  <StateLink to="^.new">
    {({ onClick }) => (
      <Button
        icon="PlusCircle"
        testId="add-locales-button"
        buttonType="primary"
        onClick={() => {
          const usage = getComputeLocalesUsageForOrganization();
          if (usage) {
            Notification.error(usage);
          } else {
            onClick();
          }
        }}>
        Add locale
      </Button>
    )}
  </StateLink>
);
AddLocaleButton.propTypes = {
  getComputeLocalesUsageForOrganization: PropTypes.func.isRequired
};

export const LocalesAdvice = props => {
  const {
    localeResource,
    insideMasterEnv,
    subscriptionPlanName,
    canChangeSpace,
    subscriptionState,
    locales,
    canCreateMultipleLocales
  } = props;

  const status = getLocalesUsageStatus({
    canCreateMultipleLocales,
    locales,
    localeResource
  });

  let advice = '';

  const upgradeLink =
    subscriptionState && subscriptionState.path ? (
      <StateLink
        to={subscriptionState.path.join('.')}
        params={subscriptionState.params}
        options={subscriptionState.options}
        className="text-link upgrade-link">
        upgrade
      </StateLink>
    ) : (
      <span>upgrade</span>
    );

  if (status === LocalesUsageStatus.MORE_THAN_ONE_LOCALE_USED || !insideMasterEnv) {
    advice = (
      <div>
        <h1 className="advice__title">
          Contentful enables publishing content in multiple languages
        </h1>
        <p className="advice__description">
          To enable localization, go to the relevant content type, open field settings, and enable
          translation for each necessary field
        </p>
        <p className="advice__description">
          After that the content editor will display multiple input fields for each locale
        </p>
      </div>
    );
  } else if (status === LocalesUsageStatus.ONE_LOCALE_USED) {
    advice = (
      <div>
        <h1 className="advice__title">
          Contentful enables publishing content in multiple languages
        </h1>
        <p className="advice__description">
          To begin translating your content, add a second locale â€“ for example,{' '}
          <strong>
            French (<code>fr-FR</code>)
          </strong>
        </p>
        <p className="advice__description">
          Note that locale settings apply space-wide: the locales that you create will affect only
          the current space
        </p>
      </div>
    );
  } else if (status === LocalesUsageStatus.LOCALES_LIMIT_REACHED) {
    advice = (
      <div>
        <h1 className="advice__title">You have reached the organization locales limit</h1>
        <p className="advice__description">
          Your current subscription plan <strong>({subscriptionPlanName})</strong> enables a maximum
          of {localeResource.limits.maximum} locales per organization
        </p>
        {canChangeSpace && (
          <p className="advice__description">
            Please {upgradeLink} if you need more locales or delete some of the existing ones
          </p>
        )}
        {!canChangeSpace && (
          <p className="advice__description">
            Please ask your organization owner to upgrade if you need more locales or delete some of
            the existing ones
          </p>
        )}
      </div>
    );
  } else if (status === LocalesUsageStatus.NO_MULTIPLE_LOCALES) {
    advice = (
      <div>
        <h1 className="advice__title">Your plan does not include multiple locales</h1>
        <p className="advice__description">
          Your current subscription plan <strong>({subscriptionPlanName})</strong> does not support
          localizing content
        </p>
        {canChangeSpace && (
          <p className="advice__description">
            Please {upgradeLink} to a plan that includes locales to benefit from this feature
          </p>
        )}
        {!canChangeSpace && (
          <p className="advice__description">
            Please ask your organization owner to upgrade to a plan that includes locales to benefit
            from this feature
          </p>
        )}
      </div>
    );
  }
  return (
    <div className="advice locale-list__advice">
      <div className="advice__frame locale-list__advice-frame" data-test-id="locales-advice">
        {advice}
      </div>
    </div>
  );
};

LocalesAdvice.propTypes = LocalesListPropTypes;

class LocalesListPricingOne extends React.Component {
  static propTypes = LocalesListPropTypes;

  render() {
    return (
      <Workbench testId="locale-list-workbench">
        <Workbench.Header
          icon={<Icon name="page-settings" scale="0.8" />}
          title={
            <>
              <Heading>Locales</Heading>
              <span className="workbench-header__kb-link">
                <KnowledgeBase target="locale" />
              </span>
            </>
          }
          actions={
            <AddLocaleButton
              getComputeLocalesUsageForOrganization={
                this.props.getComputeLocalesUsageForOrganization
              }
            />
          }
        />
        <Workbench.Content type="full">
          <LocalesTable locales={this.props.locales} />
          <LocalesAdvice {...this.props} />
        </Workbench.Content>
      </Workbench>
    );
  }
}

export default LocalesListPricingOne;
