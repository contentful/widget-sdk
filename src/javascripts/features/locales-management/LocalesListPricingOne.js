import React from 'react';
import PropTypes from 'prop-types';
import NavigationIcon from 'ui/Components/NavigationIcon';
import _ from 'lodash';
import {
  Button,
  Notification,
  Heading,
  Paragraph,
  Workbench,
} from '@contentful/forma-36-react-components';
import KnowledgeBase from 'components/shared/knowledge_base_icon/KnowledgeBase';
import { LocalesTable } from './LocalesTable';
import StateLink from 'app/common/StateLink';
import { LocalesUsageStatus, getLocalesUsageStatus } from './utils/LocalesUsageStatus';

const LocalesListPropTypes = {
  locales: PropTypes.arrayOf(PropTypes.object).isRequired,
  canCreateMultipleLocales: PropTypes.bool.isRequired,
  canChangeSpace: PropTypes.bool.isRequired,
  localeResource: PropTypes.object.isRequired,
  insideMasterEnv: PropTypes.bool.isRequired,
  subscriptionState: PropTypes.object,
  subscriptionPlanName: PropTypes.string.isRequired,
  getComputeLocalesUsageForOrganization: PropTypes.func.isRequired,
};

export const AddLocaleButton = ({ getComputeLocalesUsageForOrganization }) => (
  <StateLink path="^.new">
    {({ onClick }) => (
      <Button
        icon="PlusCircle"
        testId="add-locales-button"
        buttonType="primary"
        onClick={() => {
          const usage = getComputeLocalesUsageForOrganization();
          if (usage) {
            Notification.error(usage);
          } else {
            onClick();
          }
        }}>
        Add locale
      </Button>
    )}
  </StateLink>
);
AddLocaleButton.propTypes = {
  getComputeLocalesUsageForOrganization: PropTypes.func.isRequired,
};

export const LocalesAdvice = (props) => {
  const {
    localeResource,
    insideMasterEnv,
    subscriptionPlanName,
    canChangeSpace,
    subscriptionState,
    locales,
    canCreateMultipleLocales,
  } = props;

  const status = getLocalesUsageStatus({
    canCreateMultipleLocales,
    locales,
    localeResource,
  });

  let advice = '';

  const upgradeLink =
    subscriptionState && subscriptionState.path ? (
      <StateLink
        path={subscriptionState.path.join('.')}
        params={subscriptionState.params}
        options={subscriptionState.options}
        className="text-link upgrade-link">
        upgrade
      </StateLink>
    ) : (
      <span>upgrade</span>
    );

  if (status === LocalesUsageStatus.MORE_THAN_ONE_LOCALE_USED || !insideMasterEnv) {
    advice = (
      <div>
        <Heading className="advice__title">
          Contentful enables publishing content in multiple languages
        </Heading>
        <Paragraph className="advice__description">
          To enable localization, go to the relevant content type, open field settings, and enable
          translation for each necessary field
        </Paragraph>
        <Paragraph className="advice__description">
          After that the content editor will display multiple input fields for each locale
        </Paragraph>
      </div>
    );
  } else if (status === LocalesUsageStatus.ONE_LOCALE_USED) {
    advice = (
      <div>
        <Heading className="advice__title">
          Contentful enables publishing content in multiple languages
        </Heading>
        <Paragraph className="advice__description">
          To begin translating your content, add a second locale â€“ for example,{' '}
          <strong>
            French (<code>fr-FR</code>)
          </strong>
        </Paragraph>
        <Paragraph className="advice__description">
          Note that locale settings apply space-wide: the locales that you create will affect only
          the current space
        </Paragraph>
      </div>
    );
  } else if (status === LocalesUsageStatus.LOCALES_LIMIT_REACHED) {
    advice = (
      <div>
        <Heading className="advice__title">You have reached the organization locales limit</Heading>
        <Paragraph className="advice__description">
          Your current subscription plan <strong>({subscriptionPlanName})</strong> enables a maximum
          of {localeResource.parent.limits.maximum} locales per organization
        </Paragraph>
        {canChangeSpace && (
          <Paragraph className="advice__description">
            Please {upgradeLink} if you need more locales or delete some of the existing ones
          </Paragraph>
        )}
        {!canChangeSpace && (
          <Paragraph className="advice__description">
            Please ask your organization owner to upgrade if you need more locales or delete some of
            the existing ones
          </Paragraph>
        )}
      </div>
    );
  } else if (status === LocalesUsageStatus.NO_MULTIPLE_LOCALES) {
    advice = (
      <div>
        <Heading className="advice__title">Your plan does not include multiple locales</Heading>
        <Paragraph className="advice__description">
          Your current subscription plan <strong>({subscriptionPlanName})</strong> does not support
          localizing content
        </Paragraph>
        {canChangeSpace && (
          <Paragraph className="advice__description">
            Please {upgradeLink} to a plan that includes locales to benefit from this feature
          </Paragraph>
        )}
        {!canChangeSpace && (
          <Paragraph className="advice__description">
            Please ask your organization owner to upgrade to a plan that includes locales to benefit
            from this feature
          </Paragraph>
        )}
      </div>
    );
  }
  return (
    <div className="advice locale-list__advice">
      <div className="advice__frame locale-list__advice-frame" data-test-id="locales-advice">
        {advice}
      </div>
    </div>
  );
};

LocalesAdvice.propTypes = LocalesListPropTypes;

export class LocalesListPricingOne extends React.Component {
  static propTypes = LocalesListPropTypes;

  render() {
    return (
      <Workbench testId="locale-list-workbench">
        <Workbench.Header
          icon={<NavigationIcon icon="settings" color="green" size="large" />}
          title={
            <>
              <Heading>Locales</Heading>
              <span className="workbench-header__kb-link">
                <KnowledgeBase target="locale" />
              </span>
            </>
          }
          actions={
            <AddLocaleButton
              getComputeLocalesUsageForOrganization={
                this.props.getComputeLocalesUsageForOrganization
              }
            />
          }
        />
        <Workbench.Content type="full">
          <LocalesTable locales={this.props.locales} />
          <LocalesAdvice {...this.props} />
        </Workbench.Content>
      </Workbench>
    );
  }
}
