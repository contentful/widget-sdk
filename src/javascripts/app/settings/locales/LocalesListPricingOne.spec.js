import React from 'react';
import Enzyme from 'enzyme';
import $stateMocked from '$state';
import { Notification } from '@contentful/ui-component-library';
import LocalesListPricingOne, { AddLocaleButton, LocalesAdvice } from './LocalesListPricingOne.es6';

describe('app/settings/locales/LocalesListPricingOne', () => {
  it('should match snapshot', () => {
    const wrapper = Enzyme.shallow(
      <LocalesListPricingOne
        locales={[]}
        canCreateMultipleLocales={true}
        canChangeSpace={true}
        insideMasterEnv={true}
        localeResource={{
          usage: 0,
          limits: {
            maximum: 5
          }
        }}
        subscriptionState={{}}
        subscriptionPlanName="some subscription plan name"
        getComputeLocalesUsageForOrganization={() => false}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  describe('AddLocaleButton', () => {
    it('if getComputeLocalesUsageForOrganization returns positive value than notification should be shown', () => {
      const notificationSpy = jest.spyOn(Notification, 'error').mockImplementation(() => {});
      const wrapper = Enzyme.mount(
        <AddLocaleButton getComputeLocalesUsageForOrganization={() => true} />
      );
      wrapper.find('button').simulate('click');
      expect(Notification.error).toHaveBeenCalledWith(true);
      expect($stateMocked.go).not.toHaveBeenCalled();
      notificationSpy.mockRestore();
    });

    it('if getComputeLocalesUsageForOrganization returns negative value than $state.go should be called', () => {
      const notificationSpy = jest.spyOn(Notification, 'error').mockImplementation(() => {});
      const wrapper = Enzyme.mount(
        <AddLocaleButton getComputeLocalesUsageForOrganization={() => false} />
      );
      wrapper.find('button').simulate('click');
      expect(Notification.error).not.toHaveBeenCalled();
      expect($stateMocked.go).toHaveBeenCalledWith('^.new', undefined, undefined);
      notificationSpy.mockRestore();
    });
  });

  describe('LocalesAdvice', () => {
    beforeEach(() => {
      $stateMocked.href.mockClear();
    });

    const mount = props => {
      return Enzyme.mount(
        <LocalesAdvice
          canCreateMultipleLocales={true}
          canChangeSpace={true}
          insideMasterEnv={true}
          subscriptionState={null}
          subscriptionPlanName="some subscription plan name"
          getComputeLocalesUsageForOrganization={() => false}
          {...props}
        />
      );
    };

    it('should show correct message if status is LocalesUsageStatus.ONE_LOCALE_USED', () => {
      const wrapper = mount({
        localeResource: {
          usage: 1,
          limits: {
            maximum: 5
          }
        },
        locales: [{ id: 1 }]
      });
      [
        'Contentful enables publishing content in multiple languages',
        'To begin translating your content, add a second locale â€“ for example, French (fr-FR)',
        'Note that locale settings apply space-wide: the locales that you create will affect only the current space'
      ].map(text => {
        expect(wrapper.find('[data-test-id="locales-advice"]')).toIncludeText(text);
      });
    });

    it('should show correct message if status is LocalesUsageStatus.MORE_THAN_ONE_LOCALE_USED or environment is not equal to "master"', () => {
      const wrapper = mount({
        localeResource: {
          usage: 3,
          limits: {
            maximum: 5
          }
        },
        locales: [{ id: 1 }, { id: 2 }, { id: 3 }]
      });
      const wrapper2 = mount({
        localeResource: {
          usage: 1,
          limits: {
            maximum: 5
          }
        },
        locales: [{ id: 1 }],
        insideMasterEnv: false
      });

      [
        'Contentful enables publishing content in multiple languages',
        'To enable localization, go to the relevant content type, open field settings, and enable translation for each necessary field',
        'After that the content editor will display multiple input fields for each locale'
      ].map(text => {
        expect(wrapper.find('[data-test-id="locales-advice"]')).toIncludeText(text);
        expect(wrapper2.find('[data-test-id="locales-advice"]')).toIncludeText(text);
      });
    });

    it('should show correct message if status is LocalesUsageStatus.LOCALES_LIMIT_REACHED', () => {
      const mountReachedLimit = ({ canChangeSpace }) => {
        return mount({
          localeResource: {
            usage: 2,
            limits: {
              maximum: 2
            }
          },
          locales: [{ id: 1 }, { id: 2 }],
          subscriptionState: {
            path: ['account', 'organization', 'subscription']
          },
          canChangeSpace
        });
      };

      const canChangeSpaceWrapper = mountReachedLimit({ canChangeSpace: true });
      const cannotChangeSpaceWrapper = mountReachedLimit({ canChangeSpace: false });

      [
        'You have reached the organization locales limit',
        'Your current subscription plan (some subscription plan name) enables a maximum of 2 locales per organization'
      ].map(text => {
        expect(canChangeSpaceWrapper.find('[data-test-id="locales-advice"]')).toIncludeText(text);
        expect(cannotChangeSpaceWrapper.find('[data-test-id="locales-advice"]')).toIncludeText(
          text
        );
      });
      expect(canChangeSpaceWrapper.find('[data-test-id="locales-advice"]')).toIncludeText(
        'Please upgrade if you need more locales or delete some of the existing ones'
      );
      expect(cannotChangeSpaceWrapper.find('[data-test-id="locales-advice"]')).toIncludeText(
        'Please ask your organization owner to upgrade if you need more locales or delete some of the existing ones'
      );
      expect($stateMocked.href).toHaveBeenCalledTimes(1);
      expect($stateMocked.href).toHaveBeenCalledWith(
        'account.organization.subscription',
        undefined
      );
    });

    it('should show correct message if status is LocalesUsageStatus.NO_MULTIPLE_LOCALES', () => {
      const mountNoMultipleLocales = ({ canChangeSpace }) => {
        return mount({
          canCreateMultipleLocales: false,
          localeResource: {
            usage: 2,
            limits: {
              maximum: 2
            }
          },
          locales: [{ id: 1 }, { id: 2 }],
          subscriptionState: {
            path: ['account', 'organization', 'subscription']
          },
          canChangeSpace
        });
      };
      const canCreateSpaceWrapper = mountNoMultipleLocales({ canChangeSpace: true });
      const cannotChangeSpaceWrapper = mountNoMultipleLocales({ canChangeSpace: false });
      [
        'Your plan does not include multiple locales',
        'Your current subscription plan (some subscription plan name) does not support localizing content'
      ].map(text => {
        expect(canCreateSpaceWrapper.find('[data-test-id="locales-advice"]')).toIncludeText(text);
        expect(cannotChangeSpaceWrapper.find('[data-test-id="locales-advice"]')).toIncludeText(
          text
        );
      });
      expect(canCreateSpaceWrapper.find('[data-test-id="locales-advice"]')).toIncludeText(
        'Please upgrade to a plan that includes locales to benefit from this feature'
      );
      expect(cannotChangeSpaceWrapper.find('[data-test-id="locales-advice"]')).toIncludeText(
        'Please ask your organization owner to upgrade to a plan that includes locales to benefit from this feature'
      );
      expect($stateMocked.href).toHaveBeenCalledTimes(1);
      expect($stateMocked.href).toHaveBeenCalledWith(
        'account.organization.subscription',
        undefined
      );
    });
  });
});
