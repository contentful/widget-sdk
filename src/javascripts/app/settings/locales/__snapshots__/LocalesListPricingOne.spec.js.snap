// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/settings/locales/LocalesListPricingOne should match snapshot 1`] = `
<Workbench
  className="locale-list entity-list"
>
  <Workbench.Header>
    <Workbench.Icon
      icon="page-settings"
    />
    <Workbench.Title>
      Locales
      <span
        className="workbench-header__kb-link"
      >
        <KnowledgeBase
          target="locale"
        />
      </span>
    </Workbench.Title>
    <HeaderActions>
      <AddLocaleButton
        getComputeLocalesUsageForOrganization={[Function]}
      />
    </HeaderActions>
  </Workbench.Header>
  <Workbench.Content>
    <LocalesTable
      locales={Array []}
    />
    <LocalesAdvice
      canChangeSpace={true}
      canCreateMultipleLocales={true}
      getComputeLocalesUsageForOrganization={[Function]}
      insideMasterEnv={true}
      localeResource={
        Object {
          "limits": Object {
            "maximum": 5,
          },
          "usage": 0,
        }
      }
      locales={Array []}
      subscriptionPlanName="some subscription plan name"
      subscriptionState={Object {}}
    />
  </Workbench.Content>
</Workbench>
`;
